File: tools/export_firestore_structure.py
import json
import os
import sys
from pathlib import Path

from google.cloud import firestore


def get_structure(collection_ref):
    structure = {}
    docs = collection_ref.stream()
    for doc in docs:
        subcollections = doc.reference.collections()
        structure[doc.id] = {
            "fields": sorted(doc.to_dict().keys()),
            "subcollections": {},
        }
        for sub in subcollections:
            structure[doc.id]["subcollections"][sub.id] = get_structure(sub)
    return structure


def main(output_path):
    client = firestore.Client()
    result = {}
    for collection in client.collections():
        result[collection.id] = get_structure(collection)

    output_path.write_text(json.dumps(result, indent=2), encoding="utf-8")
    print(f"✅ Struttura esportata in {output_path}")


if __name__ == "__main__":
    if "GOOGLE_APPLICATION_CREDENTIALS" not in os.environ:
        print("❌ Variabile GOOGLE_APPLICATION_CREDENTIALS non impostata.", file=sys.stderr)
        sys.exit(1)

    output_file = Path(__file__).resolve().parent.parent / "firestore_structure.json"
    main(output_file)

File: tools/exportFirestoreStructure.js
const admin = require("firebase-admin");
const fs = require("fs");
const path = require("path");

admin.initializeApp({
  credential: admin.credential.applicationDefault(),
});

const db = admin.firestore();

async function getStructure(collectionRef) {
  const snapshot = await collectionRef.get();
  const structure = {};

  for (const doc of snapshot.docs) {
    const subcollections = await doc.ref.listCollections();
    structure[doc.id] = {
      fields: Object.keys(doc.data()),
      subcollections: {},
    };
    for (const sub of subcollections) {
      structure[doc.id].subcollections[sub.id] = await getStructure(sub);
    }
  }

  return structure;
}

async function main() {
  const collections = await db.listCollections();
  const result = {};

  for (const col of collections) {
    result[col.id] = await getStructure(col);
  }

  const outputPath = path.resolve(__dirname, "../firestore_structure.json");
  fs.writeFileSync(outputPath, JSON.stringify(result, null, 2));
  console.log(`✅ Struttura esportata in ${outputPath}`);
}

main().catch((error) => {
  console.error("❌ Errore durante l'esportazione della struttura Firestore:", error);
  process.exit(1);
});

